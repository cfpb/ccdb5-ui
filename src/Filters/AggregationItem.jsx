import { connect } from 'react-redux'
import { FormattedNumber } from 'react-intl'
import PropTypes from 'prop-types'
import React from 'react'
import { toggleFilter } from '../actions/filter'

export const AggregationItem = ( { item, fieldName, active, onClick } ) => {
  const value = item.value || item.key
  const liStyle = 'layout-row m-form-field m-form-field__checkbox'
  const id = fieldName + item.key.replace( ' ', '' )

  // 2019-04-09 JMF
  // The seemingly do-nothing <label className="a-label"></label>
  // helps maintain consistent spacing with AggBranch in IE/Chromw/Firefox

  return (
        <li className={liStyle}>
            <input type="checkbox" className="flex-fixed a-checkbox"
                   aria-label={item.key}
                   checked={active}
                   id={id}
                   onChange={onClick}
            />
            <label className="a-label flex-all bucket-key body-copy"
                   htmlFor={id}>
              {value}
            </label>
            <span className="flex-fixed bucket-count">
              <FormattedNumber value={item.doc_count} />
            </span>
        </li>
  )
}

AggregationItem.propTypes = {
  active: PropTypes.bool,
  fieldName: PropTypes.string.isRequired,
  item: PropTypes.shape( {
    // eslint-disable-next-line camelcase
    doc_count: PropTypes.number.isRequired,
    key: PropTypes.string.isRequired,
    value: PropTypes.string
  } ).isRequired,
  onClick: PropTypes.func.isRequired
}

AggregationItem.defaultProps = {
  active: false
}

export const mapStateToProps = ( state, ownProps ) => {
  const field = state.query[ownProps.fieldName]
  const value = ownProps.item.key
  const active = typeof field !== 'undefined' && field.indexOf( value ) > -1
  return { active }
}

export const mapDispatchToProps = ( dispatch, ownProps ) => ( {
  onClick: () => {
    dispatch( toggleFilter( ownProps.fieldName, ownProps.item ) )
  }
} )

const AggregationItemFilter = connect(
  mapStateToProps,
  mapDispatchToProps
)( AggregationItem )

export default AggregationItemFilter
