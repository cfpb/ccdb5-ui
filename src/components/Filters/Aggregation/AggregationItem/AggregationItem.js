import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { filterPatch, SLUG_SEPARATOR } from '../../../../constants';
import { coalesce, sanitizeHtmlId } from '../../../../utils';
import { arrayEquals } from '../../../../utils/compare';
import { replaceFilters, toggleFilter } from '../../../../actions/filter';
import { getUpdatedFilters } from '../../../../utils/filters';
import { selectAggsState } from '../../../../reducers/aggs/selectors';
import { selectQueryState } from '../../../../reducers/query/selectors';

const appliedFilters = ({ fieldName, item, aggs, filters }) => {
  // We should find the parent
  // determine if the other siblings are already checked
  // check the parent only, and uncheck the rest so that the fake check
  // will take affect
  const [parentFilter, childFilter] = item.key.split(SLUG_SEPARATOR);
  /* eslint-disable no-unexpected-multiline */
  // TODO: reformat to not need the unexpected multiline.
  const subItems = aggs
    .find((agg) => agg.key === parentFilter)
    ['sub_' + fieldName + '.raw'].buckets.map((agg) => agg.key)
    .sort();
  /* eslint-enable no-unexpected-multiline */

  const parentKey = parentFilter + SLUG_SEPARATOR;
  const selectedFilters = filters
    .filter((filter) => filter.indexOf(parentKey) > -1)
    .map((filter) => filter.replace(parentKey, ''));
  selectedFilters.push(childFilter);

  selectedFilters.sort();

  if (arrayEquals(selectedFilters, subItems)) {
    // remove subitems, add parent filter
    return filters
      .filter((filter) => filter.indexOf(parentKey) === -1)
      .concat(parentFilter);
  } else {
    // just add the single filter and apply filters
    return filters.concat(item.key);
  }
};

export const AggregationItem = ({ fieldName, item }) => {
  const aggsState = useSelector(selectAggsState);
  const queryState = useSelector(selectQueryState);
  const dispatch = useDispatch();

  const aggs = coalesce(aggsState, fieldName, []);
  const filters = coalesce(queryState, fieldName, []);
  const isActive =
    filters.includes(item.key) ||
    filters.includes(item.key.split(SLUG_SEPARATOR)[0]);

  const value = item.value || item.key;
  const liStyle = 'layout-row m-form-field m-form-field--checkbox';
  const id = sanitizeHtmlId(fieldName + '-' + item.key);

  const addFilter = () => {
    const isChildItem = item.key.indexOf(SLUG_SEPARATOR) > -1;
    // cases where its issue / product
    if (isChildItem && filterPatch.includes(fieldName)) {
      const filtersToApply = appliedFilters({ fieldName, item, aggs, filters });
      dispatch(replaceFilters(fieldName, filtersToApply));
    } else {
      dispatch(toggleFilter(fieldName, item));
    }
  };

  const removeFilter = () => {
    if (filterPatch.includes(fieldName)) {
      const filterName = item.key;
      const updatedFilters = getUpdatedFilters(
        filterName,
        filters,
        aggs,
        fieldName,
      );
      dispatch(replaceFilters(fieldName, updatedFilters));
    } else {
      dispatch(toggleFilter(fieldName, item));
    }
  };

  const onChange = () => {
    if (isActive) {
      removeFilter();
    } else {
      addFilter();
    }
  };

  return (
    <li className={liStyle}>
      <input
        type="checkbox"
        className="flex-fixed a-checkbox"
        aria-label={item.key}
        disabled={item.isDisabled}
        checked={isActive}
        id={id}
        onChange={onChange}
      />
      <label className="a-label flex-all bucket-key body-copy" htmlFor={id}>
        {value}
      </label>
      <span className="flex-fixed bucket-count">
        {item.doc_count.toLocaleString()}
      </span>
    </li>
  );
};

AggregationItem.propTypes = {
  fieldName: PropTypes.string.isRequired,
  item: PropTypes.shape({
    // eslint-disable-next-line camelcase
    doc_count: PropTypes.number.isRequired,
    key: PropTypes.string.isRequired,
    value: PropTypes.string,
    isDisabled: PropTypes.bool,
  }).isRequired,
};
