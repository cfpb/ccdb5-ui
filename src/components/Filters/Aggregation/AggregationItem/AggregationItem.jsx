import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { filterPatch, SLUG_SEPARATOR } from '../../../../constants';
import { coalesce, sanitizeHtmlId } from '../../../../utils';
import { arrayEquals } from '../../../../utils/compare';
import {
  filtersReplaced,
  filterToggled,
} from '../../../../reducers/filters/filtersSlice';
import { getUpdatedFilters } from '../../../../utils/filters';
import { selectFiltersRoot } from '../../../../reducers/filters/selectors';
import { useGetAggregations } from '../../../../api/hooks/useGetAggregations';

const appliedFilters = ({ fieldName, item, aggs, filters }) => {
  // We should find the parent
  // determine if the other siblings are already checked
  // check the parent only, and uncheck the rest so that the fake check
  // will take affect
  const [parentFilter, childFilter] = item.key.split(SLUG_SEPARATOR);
  const subItems = aggs
    .find((agg) => agg.key === parentFilter)
    ['sub_' + fieldName + '.raw'].buckets.map((agg) => agg.key)
    .sort();

  const parentKey = parentFilter + SLUG_SEPARATOR;
  const selectedFilters = filters
    .filter((filter) => filter.indexOf(parentKey) > -1)
    .map((filter) => filter.replace(parentKey, ''));
  selectedFilters.push(childFilter);

  selectedFilters.sort();

  if (arrayEquals(selectedFilters, subItems)) {
    // remove subitems, add parent filter
    return filters
      .filter((filter) => filter.indexOf(parentKey) === -1)
      .concat(parentFilter);
  } else {
    // just add the single filter and apply filters
    return filters.concat(item.key);
  }
};

export const AggregationItem = ({ fieldName, item }) => {
  const { data: aggsState, isSuccess } = useGetAggregations();
  const filtersState = useSelector(selectFiltersRoot);
  const dispatch = useDispatch();
  const aggs = coalesce(aggsState, fieldName, []);
  const filters = coalesce(filtersState, fieldName, []);

  if (!isSuccess || !aggs) {
    return null;
  }

  const isActive =
    filters.includes(item.key) ||
    filters.includes(item.key.split(SLUG_SEPARATOR)[0]);

  const value = item.value || item.key;
  const liStyle = 'layout-row m-form-field m-form-field--checkbox';
  const id = sanitizeHtmlId(fieldName + '-' + item.key);

  const addFilter = () => {
    const isChildItem = item.key.indexOf(SLUG_SEPARATOR) > -1;
    // cases where its issue / product
    if (isChildItem && filterPatch.includes(fieldName)) {
      const filtersToApply = appliedFilters({ fieldName, item, aggs, filters });
      dispatch(filtersReplaced(fieldName, filtersToApply));
    } else {
      dispatch(filterToggled(fieldName, item));
    }
  };

  const removeFilter = () => {
    if (filterPatch.includes(fieldName)) {
      const filterName = item.key;
      const updatedFilters = getUpdatedFilters(
        filterName,
        filters,
        aggs,
        fieldName,
      );
      dispatch(filtersReplaced(fieldName, updatedFilters));
    } else {
      dispatch(filterToggled(fieldName, item));
    }
  };

  const onChange = () => {
    if (isActive) {
      removeFilter();
    } else {
      addFilter();
    }
  };

  return (
    <li className={liStyle}>
      <input
        type="checkbox"
        className="flex-fixed a-checkbox"
        aria-label={item.key}
        disabled={item.isDisabled}
        checked={isActive}
        id={id}
        onChange={onChange}
      />
      <label className="a-label flex-all bucket-key body-copy" htmlFor={id}>
        {value}
      </label>
      <span className="flex-fixed bucket-count">
        {item.doc_count.toLocaleString()}
      </span>
    </li>
  );
};

AggregationItem.propTypes = {
  fieldName: PropTypes.string.isRequired,
  item: PropTypes.shape({
    doc_count: PropTypes.number.isRequired,
    key: PropTypes.string.isRequired,
    value: PropTypes.string,
    isDisabled: PropTypes.bool,
  }).isRequired,
};
