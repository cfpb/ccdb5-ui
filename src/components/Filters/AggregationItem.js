import { coalesce, sanitizeHtmlId } from '../../utils';
import { filterPatch, SLUG_SEPARATOR } from '../../constants';
import { replaceFilters, toggleFilter } from '../../actions/filter';
import { arrayEquals } from '../../utils/compare';
import { connect } from 'react-redux';
import { FormattedNumber } from 'react-intl';
import { getUpdatedFilters } from '../../utils/filters';
import PropTypes from 'prop-types';
import React from 'react';

export class AggregationItem extends React.Component {
  _onChange() {
    if (this.props.isActive) {
      this.props.removeFilter(this.props);
    } else {
      this.props.addFilter(this.props);
    }
  }

  render() {
    const { isActive, item, fieldName } = this.props;
    const value = item.value || item.key;
    const liStyle = 'layout-row m-form-field m-form-field__checkbox';
    const id = sanitizeHtmlId(fieldName + '-' + item.key);
    return (
      <li className={liStyle}>
        <input
          type="checkbox"
          className="flex-fixed a-checkbox"
          aria-label={item.key}
          disabled={item.isDisabled}
          checked={isActive}
          id={id}
          onChange={() => this._onChange()}
        />
        <label className="a-label flex-all bucket-key body-copy" htmlFor={id}>
          {value}
        </label>
        <span className="flex-fixed bucket-count">
          <FormattedNumber value={item.doc_count} />
        </span>
      </li>
    );
  }
}

export const mapStateToProps = (state, ownProps) => {
  const aggs = coalesce(state.aggs, ownProps.fieldName, []);
  const filters = coalesce(state.query, ownProps.fieldName, []);
  const value = ownProps.item.key;
  const parentKey = value.split(SLUG_SEPARATOR)[0];
  const isActive = filters.includes(value) || filters.includes(parentKey);
  return {
    isActive,
    aggs,
    filters,
  };
};

// remove parent and apply current filter
export const mapDispatchToProps = (dispatch, ownProps) => ({
  addFilter: (props) => {
    const { aggs, filters } = props;
    const { fieldName, item } = ownProps;
    const value = item.key;
    const isChildItem = value.indexOf(SLUG_SEPARATOR) > -1;
    // cases where its issue / product
    if (isChildItem && filterPatch.includes(fieldName)) {
      // We should find the parent
      // determine if the other siblings are already checked
      // check the parent only, and uncheck the rest so that the fake check
      // will take affect
      const parentFilter = value.split(SLUG_SEPARATOR)[0];
      const childFilter = value.split(SLUG_SEPARATOR)[1];
      /* eslint-disable no-unexpected-multiline */
      // TODO: reformat to not need the unexpected multiline.
      const subItems = aggs
        .find((o) => o.key === parentFilter)
        ['sub_' + fieldName + '.raw'].buckets.map((o) => o.key)
        .sort();
      /* eslint-enable no-unexpected-multiline */

      const parentKey = parentFilter + SLUG_SEPARATOR;
      const selectedFilters = filters
        .filter((o) => o.indexOf(parentKey) > -1)
        .map((o) => o.replace(parentKey, ''));
      selectedFilters.push(childFilter);

      selectedFilters.sort();

      let appliedFilters;
      if (arrayEquals(selectedFilters, subItems)) {
        // remove subitems, add parent filter
        appliedFilters = filters
          .filter((o) => o.indexOf(parentKey) === -1)
          .concat(parentFilter);
      } else {
        // just add the single filter and apply filters
        appliedFilters = filters.concat(value);
      }
      dispatch(replaceFilters(fieldName, appliedFilters));
    } else {
      dispatch(toggleFilter(fieldName, item));
    }
  },
  removeFilter: (props) => {
    const { aggs, filters } = props;
    const { fieldName, item } = ownProps;
    if (filterPatch.includes(fieldName)) {
      const filterName = item.key;
      const updatedFilters = getUpdatedFilters(
        filterName,
        filters,
        aggs,
        fieldName,
      );
      dispatch(replaceFilters(fieldName, updatedFilters));
    } else {
      dispatch(toggleFilter(fieldName, item));
    }
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(AggregationItem);

AggregationItem.propTypes = {
  isActive: PropTypes.bool,
  fieldName: PropTypes.string.isRequired,
  item: PropTypes.shape({
    // eslint-disable-next-line camelcase
    doc_count: PropTypes.number.isRequired,
    key: PropTypes.string.isRequired,
    value: PropTypes.string,
    isDisabled: PropTypes.bool,
  }).isRequired,
  removeFilter: PropTypes.func.isRequired,
  addFilter: PropTypes.func.isRequired,
};

AggregationItem.defaultProps = {
  isActive: false,
};
