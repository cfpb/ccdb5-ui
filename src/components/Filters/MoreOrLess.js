/* eslint complexity: ["error", 6] */

import PropTypes from 'prop-types';
import React from 'react';

export class MoreOrLess extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasMore: props.hasMore };
    this._toggleShowMore = this._toggleShowMore.bind(this);
  }

  _buildListComponent(bucket) {
    const itemProps = this.props.perBucketProps(bucket, {
      ...this.props.listComponentProps,
      item: bucket,
      key: bucket.key,
    });

    return <this.props.listComponent {...itemProps} />;
  }

  _toggleShowMore() {
    this.setState({
      hasMore: !this.state.hasMore,
    });
  }

  render() {
    const all = this.props.options;
    const some = all.length > 5 ? all.slice(0, 5) : all;
    const remain = all.length - 5;

    return (
      <>
        <ul>
          {this.state.hasMore
            ? all.map((bucket) => this._buildListComponent(bucket))
            : some.map((bucket) => this._buildListComponent(bucket))}
        </ul>
        {remain > 0 ? (
          <div>
            <button
              className="a-btn a-btn__link more"
              onClick={this._toggleShowMore}
            >
              {this.state.hasMore
                ? `- Show ${remain} less`
                : `+ Show ${remain} more`}
            </button>
          </div>
        ) : null}
      </>
    );
  }
}

MoreOrLess.propTypes = {
  listComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object,
  ]).isRequired,
  listComponentProps: PropTypes.object,
  options: PropTypes.array.isRequired,
  perBucketProps: PropTypes.func,
  hasMore: PropTypes.bool,
};

MoreOrLess.defaultProps = {
  listComponentProps: {},
  perBucketProps: (bucket, props) => props,
  hasMore: false,
};

export default MoreOrLess;

MoreOrLess.propTypes = {
  listComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object,
  ]).isRequired,
  listComponentProps: PropTypes.object,
  options: PropTypes.array.isRequired,
  perBucketProps: PropTypes.func,
  hasMore: PropTypes.bool,
};

MoreOrLess.defaultProps = {
  listComponentProps: {},
  perBucketProps: (bucket, props) => props,
  hasMore: false,
};
