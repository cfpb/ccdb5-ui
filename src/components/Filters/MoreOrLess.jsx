/* eslint complexity: ["error", 6] */

import PropTypes from 'prop-types'
import React from 'react'

export class MoreOrLess extends React.Component {
  constructor( props ) {
    super( props )
    this.state = { showMore: props.showMore }
    this._toggleShowMore = this._toggleShowMore.bind( this )
  }

  _buildListComponent( bucket ) {
    const itemProps = this.props.perBucketProps( bucket,
      {
        ...this.props.listComponentProps,
        item: bucket,
        key: bucket.key
      }
    )

    return <this.props.listComponent {...itemProps} />
  }

  _toggleShowMore() {
    this.setState( {
      showMore: !this.state.showMore
    } )
  }

  render() {
    const all = this.props.options
    const some = all.length > 5 ? all.slice( 0, 5 ) : all
    const remain = all.length - 5

    return (
      <div>
        <ul>
        {this.state.showMore ?
          all.map( bucket => this._buildListComponent( bucket ) ) :
          some.map( bucket => this._buildListComponent( bucket ) )
        }
        </ul>
        {remain > 0 ?
          <div>
               <button className="a-btn a-btn__link more"
                       onClick={ this._toggleShowMore }>
                 { this.state.showMore ? `- Show ${ remain } less` :
                   `+ Show ${ remain } more`
                 }
                </button>
          </div> :
         null}
      </div>
    )
  }
}

MoreOrLess.propTypes = {
  listComponent: PropTypes.oneOfType( [
    PropTypes.element,
    PropTypes.func,
    PropTypes.object
  ] ).isRequired,
  listComponentProps: PropTypes.object,
  options: PropTypes.array.isRequired,
  perBucketProps: PropTypes.func,
  showMore: PropTypes.bool
}

MoreOrLess.defaultProps = {
  listComponentProps: {},
  perBucketProps: ( bucket, props ) => props,
  showMore: false
}

export default MoreOrLess
