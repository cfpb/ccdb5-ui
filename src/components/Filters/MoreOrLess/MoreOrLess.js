import PropTypes from 'prop-types';
import { createElement, useState } from 'react';
import { coalesce, sortOptions } from '../../../utils';
import { useSelector } from 'react-redux';
import { selectFiltersRoot } from '../../../reducers/filters/selectors';

export const MoreOrLess = ({
  fieldName,
  listComponent,
  options,
  perBucketProps = (bucket, props) => props,
  hasMore = false,
}) => {
  const [currentlyHasMore, setCurrentlyHasMore] = useState(hasMore);
  const filters = useSelector(selectFiltersRoot);
  const selectedFilters = coalesce(filters, fieldName, []);
  const all = sortOptions(options, selectedFilters, fieldName);
  const some = all.length > 5 ? all.slice(0, 5) : all;
  const remain = all.length - 5;

  const buildListComponent = (bucket) => {
    const itemProps = perBucketProps(bucket, {
      fieldName,
      item: bucket,
      key: bucket.key,
    });

    //using createElement because listComponent variable
    //cannot be read properly via JSX
    return createElement(listComponent, itemProps);
  };

  const toggleShowMore = () => {
    setCurrentlyHasMore(!currentlyHasMore);
  };

  return (
    <>
      <ul>
        {currentlyHasMore
          ? all.map((bucket) => buildListComponent(bucket))
          : some.map((bucket) => buildListComponent(bucket))}
      </ul>
      {remain > 0 ? (
        <div>
          <button className="a-btn a-btn--link more" onClick={toggleShowMore}>
            {currentlyHasMore
              ? `- Show ${remain} less`
              : `+ Show ${remain} more`}
          </button>
        </div>
      ) : null}
    </>
  );
};

MoreOrLess.propTypes = {
  fieldName: PropTypes.string.isRequired,
  listComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object,
  ]).isRequired,
  options: PropTypes.array.isRequired,
  perBucketProps: PropTypes.func,
  hasMore: PropTypes.bool,
};
