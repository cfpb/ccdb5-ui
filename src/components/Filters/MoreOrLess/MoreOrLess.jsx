/* eslint complexity: ["error", 6] */
import PropTypes from 'prop-types';
import { createElement, useState } from 'react';

//TODO: will make much less complex in implementation
// eslint-disable-next-line complexity
export const MoreOrLess = ({
  listComponent,
  listComponentProps = {},
  options,
  perBucketProps = (bucket, props) => props,
  hasMore = false,
}) => {
  const [currentlyHasMore, setCurrentlyHasMore] = useState(hasMore);

  const all = options;
  const some = all.length > 5 ? all.slice(0, 5) : all;
  const remain = all.length - 5;

  const buildListComponent = (bucket) => {
    const itemProps = perBucketProps(bucket, {
      ...listComponentProps,
      item: bucket,
      key: bucket.key,
    });

    //using createElement because listComponent variable
    //cannot be read properly via JSX
    return createElement(listComponent, itemProps);
  };

  const toggleShowMore = () => {
    setCurrentlyHasMore(!currentlyHasMore);
  };

  return (
    <>
      <ul>
        {currentlyHasMore
          ? all.map((bucket) => buildListComponent(bucket))
          : some.map((bucket) => buildListComponent(bucket))}
      </ul>
      {remain > 0 ? (
        <div>
          <button className="a-btn a-btn--link more" onClick={toggleShowMore}>
            {currentlyHasMore
              ? `- Show ${remain} less`
              : `+ Show ${remain} more`}
          </button>
        </div>
      ) : null}
    </>
  );
};

MoreOrLess.propTypes = {
  listComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
    PropTypes.object,
  ]).isRequired,
  listComponentProps: PropTypes.object,
  options: PropTypes.array.isRequired,
  perBucketProps: PropTypes.func,
  hasMore: PropTypes.bool,
};
