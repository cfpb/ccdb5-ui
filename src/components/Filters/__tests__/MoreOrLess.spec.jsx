import AggregationItem from '../AggregationItem'
import configureMockStore from 'redux-mock-store'
import { IntlProvider } from 'react-intl'
import MoreOrLess from '../MoreOrLess'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import thunk from 'redux-thunk'

const fixture = [
  {key: 'alpha', doc_count: 99},
  {key: 'beta', doc_count: 99},
  {key: 'gamma', doc_count: 99},
  {key: 'delta', doc_count: 99},
  {key: 'epsilon', doc_count: 99},
  {key: 'zeta', doc_count: 99},
  {key: 'eta', doc_count: 99},
  {key: 'theta', doc_count: 99}
]

function setupEnzyme(initial) {
  const props = {
    listComponent: AggregationItem,
    options: initial
  }

  const target = shallow(<MoreOrLess {...props} />)
  return {
    props,
    target
  }
}

function setupSnapshot( showMore ) {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore( middlewares )
  const store = mockStore( {
    query: {}
  } )
  const props = {
    showMore,
    listComponent: AggregationItem,
    listComponentProps: { fieldName: 'myfield'},
    options: fixture
  }

  return renderer.create(
    <Provider locale="en" store={ store }>
      <IntlProvider locale="en">
        <MoreOrLess { ...props } />
      </IntlProvider>
    </Provider>
  )
}

describe('component:MoreOrLess', () => {
  it('expects showAll to start false and toggle all bool states', () => {
    const { target } = setupEnzyme(fixture)
    // Initial state should be false
    expect(target.state().showMore).toEqual(false)
    target.instance()._toggleShowMore()
    expect(target.state().showMore).toEqual(true)
    target.instance()._toggleShowMore()
    expect(target.state().showMore).toEqual(false)
  })

  it( 'matches - Show NN less', () => {
    const res = setupSnapshot( true )
    expect( res ).toMatchSnapshot()
  })

  it( 'matches + Show NN more', () => {
    const res = setupSnapshot( false )
    expect( res ).toMatchSnapshot()
  })
})
