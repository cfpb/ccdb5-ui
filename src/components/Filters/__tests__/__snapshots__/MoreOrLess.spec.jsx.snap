// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component:MoreOrLess matches + Show NN more 1`] = `
<div>
  <ul>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="alpha"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldalpha"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldalpha"
      >
        alpha
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="beta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldbeta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldbeta"
      >
        beta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="gamma"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldgamma"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldgamma"
      >
        gamma
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="delta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfielddelta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfielddelta"
      >
        delta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="epsilon"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldepsilon"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldepsilon"
      >
        epsilon
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
  </ul>
  <div>
    <button
      className="a-btn a-btn__link more"
      onClick={[Function]}
    >
      + Show 3 more
    </button>
  </div>
</div>
`;

exports[`component:MoreOrLess matches - Show NN less 1`] = `
<div>
  <ul>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="alpha"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldalpha"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldalpha"
      >
        alpha
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="beta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldbeta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldbeta"
      >
        beta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="gamma"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldgamma"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldgamma"
      >
        gamma
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="delta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfielddelta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfielddelta"
      >
        delta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="epsilon"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldepsilon"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldepsilon"
      >
        epsilon
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="zeta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldzeta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldzeta"
      >
        zeta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="eta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldeta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldeta"
      >
        eta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="theta"
        checked={false}
        className="flex-fixed a-checkbox"
        id="myfieldtheta"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="myfieldtheta"
      >
        theta
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          99
        </span>
      </span>
    </li>
  </ul>
  <div>
    <button
      className="a-btn a-btn__link more"
      onClick={[Function]}
    >
      - Show 3 less
    </button>
  </div>
</div>
`;
